File: examples/javascript/client_swr_pattern.js
/**
 * SWR (Stale-While-Revalidate) Pattern Example
 *
 * Demonstrates using ekoDB Functions as an SWR cache:
 * - Check cache first, return if exists
 * - On cache miss, fetch from external API
 * - Store result with TTL for auto-expiration
 */

const { EkoDBClient, Stage } = require("@ekodb/ekodb-client");
require("dotenv").config();

const BASE_URL = process.env.API_BASE_URL || "http://localhost:8080";
const API_KEY = process.env.API_BASE_KEY || "a-test-api-key-from-ekodb";

function isNotFoundError(error) {
  const message = error instanceof Error ? error.message : String(error);
  return message.includes("status 404") || /not found/i.test(message);
}

/**
 * Save a function idempotently.
 *
 * The server returns HTTP 409 ("A function with label 'X' already exists.")
 * when a function with the same fixed label already exists. On that error we
 * UPDATE the existing function via PUT /api/functions/{label} (the server's
 * GET/PUT/DELETE routes accept either the encrypted ID or the label), then
 * resolve and return its encrypted ID so the rest of the example keeps working.
 * Any other error is propagated.
 */
async function saveOrUpdate(client, script) {
  try {
    return await client.saveFunction(script);
  } catch (error) {
    if (error.message && error.message.includes("already exists")) {
      await client.updateFunction(script.label, script);
      console.log(
        `ℹ️  Function '${script.label}' already existed — updated instead`,
      );
      const existing = await client.getFunction(script.label);
      return existing.id;
    }
    throw error;
  }
}

async function main() {
  const client = new EkoDBClient(BASE_URL, API_KEY);
  await client.init();
  const runSuffix = `${process.pid}_${Date.now()}`;
  const functionLabel = `fetch_api_user_js_${runSuffix}`;
  const cacheKey = `swr:user_js:${runSuffix}:1`;
  let primaryError;

  try {
    console.log("=== ekoDB SWR (Stale-While-Revalidate) Pattern ===\n");

    console.log("Step 1: Create SWR function that acts as edge cache");

    // Using jsonplaceholder.typicode.com - a reliable free API for testing
    const swrScript = {
      label: functionLabel,
      name: "Fetch User with Cache",
      description: "SWR pattern: Check cache, fetch from API if stale",
      parameters: {
        user_id: { required: true, description: "User ID to fetch" },
        ttl: {
          required: false,
          default: 300,
          description: "Cache TTL in seconds",
        },
      },
      version: "1.0",
      tags: ["swr", "user", "cache"],
      functions: [
        // Check KV cache for user data
        Stage.kvGet(`swr:user_js:${runSuffix}:{{user_id}}`),
        Stage.if(
          // KvGet returns {value: ...} on hit, {value: null} on miss
          // So we check if "value" is not null to detect cache hit
          {
            type: "Not",
            value: {
              condition: {
                type: "FieldEquals",
                value: { field: "value", value: null },
              },
            },
          },
          // Cache hit - return cached data
          [Stage.project(["value"], false)],
          // Cache miss - fetch from API and cache
          [
            Stage.httpRequest(
              "https://jsonplaceholder.typicode.com/users/{{user_id}}",
              "GET",
              { Accept: "application/json" },
            ),
            // Store in KV with 5 minute TTL
            Stage.kvSet(
              `swr:user_js:${runSuffix}:{{user_id}}`,
              "{{http_response}}",
              300,
            ),
            // Retrieve the cached data to return
            Stage.kvGet(`swr:user_js:${runSuffix}:{{user_id}}`),
            Stage.project(["value"], false),
          ],
        ),
      ],
    };

    const scriptId = await saveOrUpdate(client, swrScript);
    console.log(`✓ Created SWR script: ${swrScript.label} (${scriptId})\n`);

    console.log("Step 2: First call - Cache miss, fetches from API");
    const result1 = await client.callFunction(scriptId, {
      user_id: "1",
      ttl: 300,
    });
    console.log("Result:", JSON.stringify(result1, null, 2));
    console.log("✓ Data fetched from external API and cached\n");

    console.log("Step 3: Second call - Cache hit, instant response from ekoDB");
    const start = Date.now();
    await client.callFunction(scriptId, { user_id: "1" });
    const duration = Date.now() - start;
    console.log(`Response time: ${duration}ms (served from cache)`);
    console.log("✓ Lightning fast cache hit\n");

    console.log("=== SWR Pattern Summary ===");
    console.log("✅ Cache miss → Fetch from API → Store in ekoDB");
    console.log("✅ Cache hit → Instant response from ekoDB");
    console.log("✅ TTL handles automatic cache invalidation");
  } catch (error) {
    primaryError = error;
    throw error;
  } finally {
    console.log("🧹 Cleaning up...");
    const cleanupErrors = [];
    try {
      const saved = await client.getFunction(functionLabel);
      await client.deleteFunction(saved.id || functionLabel);
    } catch (error) {
      if (!isNotFoundError(error)) cleanupErrors.push(error);
    }
    try {
      await client.kvDelete(cacheKey);
    } catch (error) {
      if (!isNotFoundError(error)) cleanupErrors.push(error);
    }
    if (cleanupErrors.length > 0) {
      throw new AggregateError(
        primaryError === undefined
          ? cleanupErrors
          : [primaryError, ...cleanupErrors],
        "SWR example or cleanup failed",
      );
    }
    console.log("✓ Cleanup complete\n");
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: examples/python/client_swr_pattern.py
"""
SWR (Stale-While-Revalidate) Pattern Example

Demonstrates using ekoDB Functions as an SWR cache:
- Check cache first, return if exists
- On cache miss, fetch from external API
- Store result with TTL for auto-expiration
"""

import asyncio
import json
import os
import time
from datetime import datetime, timezone
from pathlib import Path

from dotenv import load_dotenv

env_path = Path(__file__).parent.parent / ".env"
load_dotenv(env_path)

BASE_URL = os.getenv("API_BASE_URL", "http://localhost:8080")
API_KEY = os.getenv("API_BASE_KEY", "a-test-api-key-from-ekodb")
FUNCTION_LABEL = "fetch_api_user_py"
KV_KEY_TEMPLATE = "api:user:py:{{user_id}}"
KV_KEY = "api:user:py:1"


def _is_already_exists_error(err):
    """Detect the server's 409 'function already exists' response."""
    msg = str(err)
    return "409" in msg or "already exists" in msg


def unwrap(value):
    while isinstance(value, dict) and set(value) == {"type", "value"}:
        value = value["value"]
    return value


def user_payload(result):
    records = result.get("records")
    if not isinstance(records, list) or len(records) != 1:
        raise AssertionError(f"expected one SWR record, got {records!r}")
    value = unwrap(records[0].get("value"))
    if (
        not isinstance(value, dict)
        or value.get("id") != 1
        or value.get("name") != "Leanne Graham"
    ):
        raise AssertionError(f"unexpected SWR user payload: {value!r}")
    return value


async def save_or_update(client, script):
    """Save a function, falling back to an update if its label already exists."""
    label = script["label"]
    try:
        return await client.save_function(script)
    except Exception as e:
        if not _is_already_exists_error(e):
            raise
        await client.update_function(label, script)
        print(f"ℹ️  Function '{label}' already existed — updated instead")
        existing = await client.get_function(label)
        function_id = existing.get("id")
        if not function_id:
            raise RuntimeError(f"Updated function '{label}' did not include an id")
        return function_id


async def main():
    from ekodb_client import Client, Stage

    client = Client.new(BASE_URL, API_KEY)

    # This SWR example uses KV operations (kv_get/kv_set) — no collection needed

    print("=== ekoDB SWR (Stale-While-Revalidate) Pattern ===\n")

    print("Step 1: Create SWR function that acts as edge cache")

    # Using jsonplaceholder.typicode.com - a reliable free API for testing
    swr_script = {
        "label": FUNCTION_LABEL,
        "name": "Fetch User with Cache",
        "description": "SWR pattern: Check cache, fetch from API if stale",
        "parameters": {
            "user_id": {"required": True, "description": "User ID to fetch"},
            "ttl": {
                "required": False,
                "default": 300,
                "description": "Cache TTL in seconds",
            },
            "cached_at": {
                "required": True,
                "description": "UTC timestamp when cached",
            },
        },
        "version": "1.0",
        "tags": ["swr", "user", "cache"],
        "functions": [
            # Check KV cache for user data
            Stage.kv_get(KV_KEY_TEMPLATE),
            Stage.if_condition(
                # KvGet returns {value: ...} on hit, {value: null} on miss
                # So we check if "value" is not null to detect cache hit
                {
                    "type": "Not",
                    "value": {
                        "condition": {
                            "type": "FieldEquals",
                            "value": {"field": "value", "value": None},
                        }
                    },
                },
                # Cache hit - return cached data
                [Stage.project(["value"], False)],
                # Cache miss - fetch from API and cache
                [
                    Stage.http_request(
                        "https://jsonplaceholder.typicode.com/users/{{user_id}}",
                        "GET",
                        {"Accept": "application/json"},
                    ),
                    # Store in KV with 5 minute TTL
                    Stage.kv_set(
                        KV_KEY_TEMPLATE,
                        "{{http_response}}",
                        300,
                    ),
                    # Retrieve the cached data to return
                    Stage.kv_get(KV_KEY_TEMPLATE),
                    Stage.project(["value"], False),
                ],
            ),
        ],
    }

    if await client.kv_exists(KV_KEY):
        await client.kv_delete(KV_KEY)

    script_id = None
    operation_error = None
    try:
        script_id = await save_or_update(client, swr_script)
        print(f"✓ Created SWR script: {swr_script['label']} ({script_id})\n")

        print("Step 2: First call - Cache miss, fetches from API")
        result1 = await client.call_function(
            FUNCTION_LABEL,
            {
                "user_id": "1",
                "ttl": 300,
                "cached_at": datetime.now(timezone.utc).isoformat(),
            },
        )
        print(f"Result: {json.dumps(result1, indent=2)}")
        payload1 = user_payload(result1)
        print("✓ Data fetched from external API and cached\n")

        print("Step 3: Second call - Cache hit, instant response from ekoDB")
        start = time.time()
        result2 = await client.call_function(
            FUNCTION_LABEL,
            {"user_id": "1", "cached_at": datetime.now(timezone.utc).isoformat()},
        )
        duration = (time.time() - start) * 1000
        print(f"Response time: {duration:.0f}ms (served from cache)")
        print(f"Result: {json.dumps(result2, indent=2)}")
        payload2 = user_payload(result2)
        if payload2 != payload1:
            raise AssertionError("cache hit returned different user data")
        print("✓ Lightning fast cache hit\n")
    except BaseException as error:  # noqa: BLE001 - cleanup must run on cancellation
        operation_error = error

    # Cleanup
    print("🧹 Cleaning up...")
    cleanup_errors = []
    if script_id is not None:
        try:
            await client.delete_function(script_id)
        except Exception as error:  # noqa: BLE001 - attempt every cleanup
            cleanup_errors.append(f"function {script_id}: {error}")
    try:
        if await client.kv_exists(KV_KEY):
            await client.kv_delete(KV_KEY)
    except Exception as error:  # noqa: BLE001 - preserve the operation error
        cleanup_errors.append(f"KV key {KV_KEY}: {error}")

    if operation_error is not None:
        if cleanup_errors:
            operation_error.add_note(
                "cleanup also failed: " + "; ".join(cleanup_errors)
            )
            print("⚠️  Cleanup errors: " + "; ".join(cleanup_errors))
        raise operation_error
    if cleanup_errors:
        raise RuntimeError("Cleanup failed: " + "; ".join(cleanup_errors))
    print("✓ Cleanup complete\n")

    print("=== SWR Pattern Summary ===")
    print("✅ Cache miss → Fetch from API → Store in ekoDB")
    print("✅ Cache hit → Instant response from ekoDB")
    print("✅ TTL handles automatic cache invalidation")


if __name__ == "__main__":
    asyncio.run(main())
